### PR TITLE
Added extracting of templated params from `params`

### LINKED ISSUE #34: Generate types for `params()` automatically by TS string parsing
Backport similar feature from Nano Stores Router https://github.com/nanostores/router/pull/12

### PR DESCRIPTION
**Pull request by BEGEMOT9I**
Close #34 

Hi! This pull request solves the problem of autoinferring parameters for the `params` function, but with two edge cases:
1. If nested templates are used (see example in test file), we can't be sure which of the parameter names will be used. Therefore, for this case, it remains possible to set parameters in the function.
2. The case of using functions as parameters, for example, the `count` function, is not processed here. This requires an extension of the return type signature of the `TranslationFunction` function (even if it will be readable), and will likely require updating the types of all utilities.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
   ],
   "scripts": {
     "unit": "tsm node_modules/uvu/bin.js . '\\.test\\.(ts|js)$'",
-    "test": "c8 yarn unit && eslint . && check-dts && size-limit",
+    "test": "c8 yarn unit && eslint . --report-unused-disable-directives && check-dts && size-limit",
```
Comment (ai) [inline package.json:18]: This change should be in different PR

File: params/index.d.ts [hunk 2]
```diff
@@ -24,6 +30,12 @@ interface Params {
    * @param input Template string.
    * @return Transform for translation.
    */
+  <Input extends string>(input: Input): TranslationFunction<
```
Comment (ai) [inline params/index.d.ts:33]: In Router we were forced to define argument as `const` to be sure that TS will use `"/posts/:id"` type instead of just `string` https://github.com/nanostores/router/blob/main/index.d.ts#L132C30-L132C35

Why we don’t need it here?

File: params/index.d.ts [hunk 3]
```diff
@@ -3,6 +3,12 @@ import type {
...
 
+type ExtractTemplateParams<Str extends string> =
+  Str extends `${infer Pre}{${infer Param}}${infer Post}`
+    ? { [key in Param]: string | number } & ExtractTemplateParams<Pre> &
+        ExtractTemplateParams<Post>
+    : {}
+
 interface Params {
   /**
    * Add `{name}` parameters to translation strings.
```
Comment (ai) [inline params/index.d.ts:14]: We will need also to change docs in TypeDoc example and in `README.md`

File: params/index.d.ts (hunk 2 above)
Comment (BEGEMOT9I) [reply params/index.d.ts:33]: Bcs in router types it's needed to narrow types for the other consumers, that will be working with returned type. Here I don't see cases why it might be necessary, but I might be wrong)

**Comment by BEGEMOT9I** [comment]
@ai wdyt about adding a case with `params` and a `count` function to the demo?

**Comment by ai** [comment]
Thanks. Released at 0.11.

File: params/index.d.ts (hunk 2 above)
Comment (milushov) [reply params/index.d.ts:33]: I was passing by and wanted to give a hint that using canonical URLs will be safer in the future. Just press `y` and the URL will become this: https://github.com/nanostores/router/blob/e927e23d509d1cde4140653071743a3c370de70b/index.d.ts#L132C30-L132C35
